### ISSUE TITLE
How to define a trip in SIRI?

### ISSUE THREAD

**Issue by EyalBerger** [hasadna/open-bus#174]
# How to define a trip in SIRI?

The most basic question for SIRI data analyses is how to define and index a single trip.

## First option:  "trip_id_to_date"

The column "trip_id_to_date" in SIRI might seem like a good choice, but actually different trips (in a same day or in different days) can share the same "trip_id_to_date" (*example will be added soon*).

Due to the absent of an original index in SIRI, we will need to build our own trip index. let start. 

## Second option:  "trip_id_to_date": route_id" + "planned_start_time" + date

We can build a trip index based on 3 components:
* route index = column "route_id" in SIRI
* trip start time (planned) = column "planned_start_time" in SIRI 
* trip date (planned)

Trip date can be calculated from "timestamp" column but needs some manipulations to overcome some cases when the trip starts close to midnight. One method is to add day_cutoff parameter as described in this issue - #103.   

@cjer @AvivSela - does SIRI raw data contains the date of "planned_start_time", and we can add it as a new column to our database?

**Comment by AvivSela** [comment]
I think we need to find a way to define trip in both GTFS and SIRI - eventually we will need to combine them together. 

The most static definition of route (in GTFS) is **route_desc**: 
> This number accompanies the line throughout its life and does not change, and constitutes a kind of identity card for the line.

What do you say about the following trip-id format: {route_desc}_{yyyymmddThhmm}.
For example: 11900-2-10_20190411T2124.

The pros: 
- We could skip all the challenges regarding the indexes that are relevant just for the current date. 
- We could have share trip index in GTFS and SIRI

The cons:
- We wont able to calculate trip id with only SIRI data.
- We might need to find way to convert some chars in route_desc. For example:
  -   '11164-2-ק',
  -   '33011-2-0',
  -   '10571-1-ב',
  -   '10827-2-#'

**Comment by EyalBerger** [comment]
Sounds good.

but...

1. What is **route_desc**?  I looked over a route stats file, but didn't find any field with this topic. What is the difference between this field and **route_id** or **route_mkt**? I also didn't find it on a SIRI file.

2.  We still need to overcome the "SIRI to date" challenge that I mentioned above. 

3. GTFS and SIRI doesn't contains the same trips (for specific route_id and date), so we would need to be able to calculate trip id with only SIRI data. Basically, we can index both SIRI and GTFS in the same way based on their own fields, so we would have also trips that are unique to SIRI/GTFS.

**Comment by cjer** [comment]
Just to answer your first question: in `gtfs_stats` I split `route_desc` into three fields: `route_mkt`, `route_direction`, `route_alternative`. 
I'll reply with more of my thoughts, and the rest of the bigger context we talked about, later on today or tomorrow

**Comment by EyalBerger** [comment]
Great, I will try to build a trip index using this fields. 

We also agreed that the index must be represented as an Integer.

Regrading trip date, I have opened a separate issue num. #182

**Comment by AvivSela** [comment]
@EyalBerger, We agreed that that the index will be combination of `route_mkt`, `route_direction`, `route_alternative`, `trip departure timestamp`

**Comment by EyalBerger** [comment]
Yes. When you write "trip departure timestamp" you mean "planned_start_time" hour and "planned_start_time" date, right?

**Comment by AvivSela** [comment]
Let's change the term SIRI trip into SIRI ride. The definition of SIRI ride is: **Group of points that represent a GTFS route**. It is not necessary that the SIRI ride contain a reference to specific GTFS trip.

**Comment by EyalBerger** [comment]
sounds fine.
According to this definition, what are we planning to do with "trips" with partial points data that can't be matched to a specific GTFS route? we still need to name them and mark them as incomplete "ride"/"trip".

**Comment by EyalBerger** [comment]
Question about calculating **ride_index** : 
Does `route_desc` (or `route_mkt, route_direction & route_alternative`) are available in SIRI raw data? or we have only `route_id` in SIRI?

**Comment by cjer** [comment]
None of route_desc fields are in SIRI, only route_id

**Comment by EyalBerger** [comment]
We agreed to add bus_id to the index.

**Comment by EyalBerger** [comment]
@AvivSela suggested another way to define rides in SIRI: 
Run on bus_id  column - first ride starts from the first `time recorded` per bus_id in the data and last until the first time there is no GPS coordinates data, which means a new ride had started. and so on. 

I will check both definitions and compare them.

**Comment by EyalBerger** [comment]
Just some small updates:
I decided to do this analysis a firm by firm, due to the differences in reporting conventions, starting from Egged Taavura (agency id=4).

As a benchmark, I created two simple trip indexes: 
* Route_id + date + trip_id_to_date + bus_id
* Route_id + date + planned_start_time + bus_id 

I was happy to found out that they are 100% correlated, when checking all the SIRI reports of the firm for a single day.

I also developed two more trip indexes using the GPS coordinates order per bus id.
They both failed to "beat" the simple indexses, when testing some examples on a map. They failed mainly because of "garbage" rows (with or without GPS coordinates) which "confused" them. 
Back to the simple indexes, my next step is to check them versus the GTFS data. Then, I will extend my analyze to other firms.
